plutus-core/plutus-ir/src/PlutusIR/Transform/RecSplit.hs

```diff
@@ -0,0 +1,120 @@
+
+Currently the implementation relies on 'Unique's, so there is the assumption of global uniqueness of the input term.
+However, the algorithm could be changed to work without this assumption (has not been tested).
+-}
+
+{-|
+Apply letrec splitting, recursively in bottom-up fashion.
+-}
+recSplit :: forall uni fun a name tyname.
+           (Ord name, Ord tyname, PLC.HasUnique tyname PLC.TypeUnique,
+            PLC.HasUnique name PLC.TermUnique)
+         => Term tyname name uni fun a
+         -> Term tyname name uni fun a
+recSplit = transformOf termSubterms recSplitStep
+  where
+    recSplitStep :: Term tyname name uni fun a -> Term tyname name uni fun a
+    recSplitStep = \case
+        -- See Note [LetRec splitting pass]
+        Let a Rec bs t ->
+            let
+                -- a table from principal id to the its corresponding whole let-binding
+                bindingsTable :: M.Map PLC.Unique (Binding tyname name uni fun a)
+                bindingsTable = M.fromList . NE.toList $ fmap (\ b -> (principal b, b)) bs
+                hereSccs =
+                           -- TODO: use error infrastructure
+                           fromRight (error "Cycle detected in the scc-graph. This shouldn't happen in the first place.")
+                           -- we take the topological sort (for the correct order)
+                           -- from the SCCs (for the correct grouping) of the local dep-graph
+                           . AM.topSort . AM.scc $ buildLocalDepGraph bs
+            in foldl' (\ acc scc ->
+                          Let a
+                          (if AM.isAcyclic $ AMN.fromNonEmpty scc then NonRec else Rec)
+                          (NE.fromList . M.elems . M.restrictKeys bindingsTable $ AMN.vertexSet scc)
+                          acc) t hereSccs
+        t  -> t
+
+
+{-|
+It constructs a dependency graph between the introduced bindings of the currently-examined let,
+and takes
+-}
+buildLocalDepGraph :: forall uni fun a name tyname.
+                     (Ord name, Ord tyname, PLC.HasUnique tyname PLC.TypeUnique, PLC.HasUnique name PLC.TermUnique)
+                   => NE.NonEmpty (Binding tyname name uni fun a) -> AM.AdjacencyMap PLC.Unique
+buildLocalDepGraph bs = AM.overlays . NE.toList $ fmap bindingSubGraph bs
+    where
+      -- a map of a all introduced binding ids to their belonging principal id
+      idTable :: M.Map PLC.Unique PLC.Unique
+      idTable = foldMap1 (\ b -> M.fromList (fmap (,principal b) $ b^..bindingIds)) bs
+
+      bindingSubGraph :: Binding tyname name uni fun a -> AM.AdjacencyMap PLC.Unique
+      bindingSubGraph b =
+          let freeUniques = S.map (^.PLC.theUnique) (fvBinding b)
+                            -- why NonRec: if it is a datatype-binding we treat it as non-recursive
+                            -- to reveal if it depends on itself,
+                            -- i.e. the typeconstructor is free at the domain-part of any of its dataconstructor types
```

Comment:
This is a bit hard to understand, and I think it's a key point. Possibly worth a proper note.